optuna/study/_tell.py

```diff
@@ -0,0 +1,221 @@
+    """
+
+    if not isinstance(trial, (trial_module.Trial, int)):
+        raise TypeError("Trial must be a trial object or trial number.")
+
+    if state == TrialState.COMPLETE:
+        if values is None:
+            raise ValueError(
+                "No values were told. Values are required when state is TrialState.COMPLETE."
+            )
+    elif state in (TrialState.PRUNED, TrialState.FAIL):
+        if values is not None:
+            raise ValueError(
+                "Values were told. Values cannot be specified when state is "
+                "TrialState.PRUNED or TrialState.FAIL."
+            )
+    elif state is not None:
+        raise ValueError(f"Cannot tell with state {state}.")
+
+    if isinstance(trial, trial_module.Trial):
+        trial_number = trial.number
+        trial_id = trial._trial_id
+    elif isinstance(trial, int):
+        trial_number = trial
+        try:
+            trial_id = study._storage.get_trial_id_from_study_id_trial_number(
+                study._study_id, trial_number
+            )
+        except NotImplementedError as e:
+            warnings.warn(
+                "Study.tell may be slow because the trial was represented by its number but "
+                f"the storage {study._storage.__class__.__name__} does not implement the "
+                "method required to map numbers back. Please provide the trial object "
+                "to avoid performance degradation."
+            )
+
+            trials = study.get_trials(deepcopy=False)
+
+            if len(trials) <= trial_number:
+                raise ValueError(
+                    f"Cannot tell for trial with number {trial_number} since it has not been "
+                    "created."
+                ) from e
+
+            trial_id = trials[trial_number]._trial_id
+        except KeyError as e:
+            raise ValueError(
+                f"Cannot tell for trial with number {trial_number} since it has not been "
+                "created."
+            ) from e
+    else:
+        assert False, "Should not reach."
+
+    frozen_trial = study._storage.get_trial(trial_id)
+    warning_message = None
+
+    if frozen_trial.state.is_finished() and skip_if_finished:
+        _logger.info(
+            f"Skipped telling trial {trial_number} with values "
+            f"{values} and state {state} since trial was already finished. "
+            f"Finished trial has values {frozen_trial.values} and state {frozen_trial.state}."
+        )
+        return copy.deepcopy(frozen_trial)
+
+    if state is None and values is None:
+        state = TrialState.FAIL
+        message_values_and_state_not_specified = (
+            "Study.tell is called with `values` and `state` set to None but "
+            "you have to specify either `values` or `state` at least. "
+            "`values` is required if a trial finishes successfully or "
+            "`state` is required if a trial is fails or pruned."
+        )
```

Comment:
Not sure how to fix it but this message is displayed to users calling `Study.optimize` returning `None` from their objective functions, which might be less intuitive given the message is specific to `Study.tell`. In `master`, the message in this case is `Trial 0 failed, because the value None could not be cast to float.`. I think we want to keep this behavior. What do you think?

Reply:
Thank you for your consideration, your feedback makes sense. I'll try to revise my PR.
I'm considering that it changes the error message on the value of `suppress_warning` assuming it be `True` only when `_tell_with_warning` is invoked by `study._run_trial`.

Reply:
@hvy 
I tried to keep the current behavior as much as possible. b2aff2b drops a `raise` and checks a value even if it is `None`. After this change, `Study.optimize` make a trial whose value is `None` `TrialState.FAIL` with the message `Trial 0 failed, because the value None could not be cast to float.`. Although this patch changes the behavior when `Study.tell` is called directly, I think it doesn't change in the perspective of `Study.optimize` as `Study.optimize` already checked a value in `_run_trial` in the existing implementation. I'd like to hear your feedback again.